ucp/application_context.py

```diff
@@ -0,0 +1,376 @@
+from ._libs import ucx_api
+from .exceptions import UCXCanceled, UCXError, UCXWarning
+
+
+def asyncio_handle_exception(loop, context):
+    msg = context.get("exception", context["message"])
+    if isinstance(msg, UCXCanceled):
+        log = logging.debug
+    elif isinstance(msg, UCXWarning):
+        log = logging.warning
+    else:
+        log = logging.error
+    log("Ignored except: %s %s" % (type(msg), msg))
+
+
+async def exchange_peer_info(ucp_endpoint, msg_tag, ctrl_tag, guarantee_msg_order):
+    """Help function that exchange endpoint information"""
+
+    my_info = struct.pack("QQi", msg_tag, ctrl_tag, guarantee_msg_order)
+    peer_info = bytearray(len(my_info))
+
+    await asyncio.gather(
+        send_recv.stream_recv(ucp_endpoint, peer_info, len(peer_info)),
+        send_recv.stream_send(ucp_endpoint, my_info, len(my_info)),
+    )
+    peer_msg_tag, peer_ctrl_tag, peer_guarantee_msg_order = struct.unpack(
+        "QQi", peer_info
+    )
+
+    if peer_guarantee_msg_order != guarantee_msg_order:
+        raise ValueError("Both peers must set guarantee_msg_order identically")
+
+    return {
+        "msg_tag": peer_msg_tag,
+        "ctrl_tag": peer_ctrl_tag,
+        "guarantee_msg_order": peer_guarantee_msg_order,
+    }
+
+
+class CtrlMsg:
+    """Serialization and deserialization of a control message
+
+    For now we have one opcode `1` which means shutdown.
+    The opcode takes `close_after_n_recv`, which is the number of
+    messages to receive before the worker should close.
+    """
+
+    fmt = "QQ"
+    nbytes = struct.calcsize(fmt)
+
+    @staticmethod
+    def serialize(opcode, close_after_n_recv):
+        return struct.pack(CtrlMsg.fmt, opcode, close_after_n_recv)
+
+    @staticmethod
+    def deserialize(serialized_bytes):
+        return struct.unpack(CtrlMsg.fmt, serialized_bytes)
+
+
+def handle_ctrl_msg(ep_weakref, log, msg, future):
+    """Function that is called when receiving the control message"""
+    try:
+        future.result()
+    except UCXCanceled:
+        return  # The ctrl signal was canceled
+    logging.debug(log)
+    ep = ep_weakref()
+    if ep is None or ep.closed():
+        return  # The endpoint is closed
+
+    opcode, close_after_n_recv = CtrlMsg.deserialize(msg)
+    if opcode == 1:
+        ep.close_after_n_recv(close_after_n_recv, count_from_ep_creation=True)
+    else:
+        raise UCXError("Received unknown control opcode: %s" % opcode)
+
+
+def setup_ctrl_recv(ep):
+    """Help function to setup the receive of the control message"""
+
+    msg = bytearray(CtrlMsg.nbytes)
+    log = "[Recv shutdown] ep: %s, tag: %s" % (hex(ep.uid), hex(ep._ctrl_tag_recv))
+    ep.pending_msg_list.append({"log": log})
+    shutdown_fut = send_recv.tag_recv(
+        ep._worker,
+        msg,
+        len(msg),
+        ep._ctrl_tag_recv,
+        pending_msg=ep.pending_msg_list[-1],
+    )
+
+    shutdown_fut.add_done_callback(partial(handle_ctrl_msg, weakref.ref(ep), log, msg))
```

Comment:
Why do we use a `weakref` here?

Reply:
To avoid holding a reference to ApplicationContext, which would make ucp.reset() fails.

Reply:
Got it. Thanks!